actors/builtin/market/market_test.go

```diff
@@ -274,27 +271,119 @@ func TestMarketActor(t *testing.T) {
+
+			actor.withdrawClientBalanceOK(rt, client, withDrawAmt, actualWithdrawn)
 		})
 
-		// TODO: withdraws limited by slashing
-		// TODO: withdraws limited by locked balance
-		// https://github.com/filecoin-project/specs-actors/issues/465
+		t.Run("worker balance after withdrawal must ALWAYS be greater than or equal to locked amount", func(t *testing.T) {
+			rt, actor := setup()
+
+			// create the deal to publish and the epochs around it
+			startEpoch := abi.ChainEpoch(10)
+			endEpoch := abi.ChainEpoch(20)
+			currentEpoch := abi.ChainEpoch(5)
+			deal := actor.generateUnVerifiedDealProposal(client, provider, startEpoch, endEpoch)
+
+			// ensure client and provider have enough funds to lock for the deal
+			actor.addParticipantFunds(rt, client, deal.ClientBalanceRequirement())
+			actor.addProviderFunds(rt, provider, owner, worker, deal.ProviderBalanceRequirement())
+
+			// publish the deal so that provider collateral is locked
+			rt.SetEpoch(currentEpoch)
+			actor.publishDealOK(rt, deal, owner, worker, provider)
+			rt.GetState(&st)
+			require.Equal(t, deal.ProviderCollateral, st.GetLockedBalance(rt, provider))
+
+			// provider cannot withdraw any funds since all it's balance is locked
+			withDrawAmt := abi.NewTokenAmount(1)
+			actualWithdrawn := abi.NewTokenAmount(0)
+			actor.withdrawStorageMinerBalanceOK(rt, owner, worker, provider, withDrawAmt, actualWithdrawn)
+
+			// add some more funds to the provider & ensure withdrawal is limited by the locked funds
+			actor.addProviderFunds(rt, provider, owner, worker, abi.NewTokenAmount(25))
+			withDrawAmt = abi.NewTokenAmount(30)
+			actualWithdrawn = abi.NewTokenAmount(25)
+
+			actor.withdrawStorageMinerBalanceOK(rt, owner, worker, provider, withDrawAmt, actualWithdrawn)
+		})
+
+		t.Run("worker balance after withdrawal must account for slashed funds", func(t *testing.T) {
+			rt, actor := setup()
+
+			// create the deal to publish and the epochs around it
+			startEpoch := abi.ChainEpoch(10)
+			endEpoch := abi.ChainEpoch(20)
+			currentEpoch := abi.ChainEpoch(5)
+			deal := actor.generateUnVerifiedDealProposal(client, provider, startEpoch, endEpoch)
+
+			// ensure client and provider have enough funds to lock for the deal
+			actor.addParticipantFunds(rt, client, deal.ClientBalanceRequirement())
+			actor.addProviderFunds(rt, provider, owner, worker, deal.ProviderBalanceRequirement())
+
+			// publish the deal
+			rt.SetEpoch(currentEpoch)
+			dealID := actor.publishDealOK(rt, deal, owner, worker, provider)
+
+			// activate the deal
+			actor.activeDealOK(rt, dealID, endEpoch + 1, provider)
+			st :=actor.mustGetDealState(rt, dealID)
+			require.EqualValues(t, currentEpoch, st.SectorStartEpoch)
+
+			// slash the deal
+			rt.SetEpoch(currentEpoch+1)
+			actor.terminateDealOK(rt, dealID, provider)
+			st = actor.mustGetDealState(rt, dealID)
+			require.EqualValues(t, currentEpoch+1, st.SlashEpoch)
+
+			// provider cannot withdraw any funds since all it's balance is locked
```

Comment:
We need more tests later than when the deal state is updated by cron, the miner can withdraw the payment for the epochs during which the deal was honoured, and that the client only recovers the amount for the rest. But this can wait for your re-work of slashing.